File: src/processing/balances.py
"""Account balance and net worth calculations."""

import pandas as pd


def compute_account_balances(df: pd.DataFrame) -> pd.DataFrame:
    """
    Returns a DataFrame with a running balance per account.
    Columns: Period, Account, Amount, Income/Expense, signed_amount, balance
    """
    df = df.copy()

    def sign(row):
        t = row["Income/Expense"]
        if t == "Income":
            return row["Amount"]
        elif t == "Expense":
            return -row["Amount"]
        elif t == "Transfer-In":
            # Money arriving into the account listed in 'Account'
            return row["Amount"]
        elif t == "Transfer-Out":
            # Money leaving the account listed in 'Account'
            return -row["Amount"]
        elif t == "Transfer":
            return 0.0
        elif t == "Saving":
            return -row["Amount"]
        elif t == "Income Balance":
            # Reconciliation adjustment bringing tracked balance up to actual.
            return row["Amount"]
        elif t == "Expense Balance":
            # Reconciliation adjustment bringing tracked balance down to actual.
            return -row["Amount"]
        return 0.0

    df["signed_amount"] = df.apply(sign, axis=1)
    df = df.sort_values("Period")
    df["balance"] = df.groupby("Account")["signed_amount"].cumsum()
    return df

File: web/src/lib/balances.ts
// Per-account running balances — a direct port of src/processing/balances.py.
// This is where transfer semantics matter: a Transfer moves money between
// accounts (In +, Out −) but nets to zero across the ledger, so net worth is
// unchanged while each account's balance shifts.
import type { Txn, TxnType } from '../db'

// Signed contribution of a row to its account's balance (balances.py `sign`).
export function signedAmount(type: TxnType, amount: number): number {
  switch (type) {
    case 'Income':
    case 'Transfer-In':
    case 'Adjustment-In':
      return amount
    case 'Expense':
    case 'Transfer-Out':
    case 'Saving':
    case 'Adjustment-Out':
      return -amount
    default:
      return 0
  }
}

// True for money genuinely entering/leaving net worth (excludes transfers, which
// only move between the user's own accounts).
export function affectsNetWorth(type: TxnType): boolean {
  return type !== 'Transfer-In' && type !== 'Transfer-Out'
}

// Closing balance per account across the whole ledger.
export function accountBalances(txns: Txn[]): Record<string, number> {
  const out: Record<string, number> = {}
  for (const t of txns) {
    out[t.account] = (out[t.account] ?? 0) + signedAmount(t.type, t.amount)
  }
  return out
}
